[SE-0293] Implement revision #3 of property wrapper parameters

hborla opened the pull request:
* The distinction between API wrappers and implementation-detail wrappers is formalized, and determined by the compiler based on whether the property wrapper type defines `init(projectedValue:)` (and also defines `var projectedValue`).
* Only API property wrappers on parameters use caller-side application of the property wrapper, and are part of the function signature. Implementation-detail property wrappers on parameters use callee-side application of the property wrapper, and have no external effect on the function.

hborla committed:
[Property Wrappers] Implement implementation-detail property wrappers for
parameters.

slavapestov commented on include/swift/AST/PropertyWrappers.h:
```diff
@@ -170,16 +170,36 @@ void simple_display(llvm::raw_ostream &os, PropertyWrapperLValueness l);
+
+  explicit operator bool() const { return isValid(); }
+
+  friend bool operator==(const PropertyWrapperAuxiliaryVariables &lhs,
```
We no longer use the AnyValue type eraser to store cached request values so unless you need it for some other purpose, you don’t have to overload operator== anymore.

hborla replied:
I removed it, thanks!

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please build toolchain macOS platform

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[Property Wrappers] Correct the source location of applied property
wrapper expressions in property wrapper parameter thunks, and add
SILGen tests for property wrapper parameters in inlinable functions.

hborla committed:
[Property Wrappers] Fixes and tests for inaccessible and unavailable
parameter wrappers.

hborla commented:
I'm going to split the request refactoring out into a separate PR

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

tkremenek commented:
@swift-ci clean test

hborla committed:
[NFC] Add SE-0293 to CHANGELOG.md

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test Windows platform
